wordpress-mobile/gutenberg-mobile
# Additional alignment of focus and selection logic with the web's

Fixes #1029 

Gutenberg PR: https://github.com/WordPress/gutenberg/pull/15999

This PR makes changes to the Aztec wrapper (Android) as well as the RN app to bring it closer to the web as far as focus and selection logic goes.

### Important changes:

1. Try to avoid having Aztec (at the wrapper level and down) request focus automatically. We rely on the JS logic to do that.
  1. `setFocusableInTouchMode`, `setFocusable` is set to `false` until Aztec needs to receive focus. Also setting it to false when Aztec loses focus.
  2. Completely removed the `isSelected()` prop since the component should not use it as a flag to request focus

2. Schedule a `requestRectangleOnScreen` to scroll into view
When creating a new RichText based block, Aztec receives focus OK but the container doesn't seem to get scrolled to make sure the edit box is in the viewport. https://github.com/wordpress-mobile/gutenberg-mobile/pull/1076/files#diff-5693404c4c8f8d757cd298cc5aabc561R185 asks the system to bring the caret into the viewport. In a future iteration of this logic, we should try to bring the whole _block_ view into the viewport.

3. Schedule the `showSoftKeyboard` to happen on a next frame

4. Remove the 500ms delay for onContentSizeChange and just schedule it for next frame. This will actually dramatically improve the performance perception of the editor on Android, making the writing flow smoother. There's still the visible glitch where blocks start with minimum height and then expand though and fixing that (separate PR) will also greatly benefit the performance.

5. No need to try blurring other components in `block-holder`'s onFocus. Component will self-blur via https://github.com/WordPress/gutenberg/pull/15999/files#diff-4828a21853e899e5a36faecfa96d83e8R753

6. Align block-holder's `withDispatch` with the web's https://github.com/wordpress-mobile/gutenberg-mobile/pull/1076/files#diff-2004b9687d71de0f0a9816269bb4a6a2R276


To test:
1. Run the demo app and place the caret in the middle (or the end for that matter) of a paragraph block
2. Tap "Enter" to split the block into two paragraph blocks
3. Notice that the virtual keyboard is still visible and the caret is normally placed at the beginning of the newly created block.

Also, test focus and selection in general, it should work as expected.

Known issue: at least on Android v7.1.2, sometimes a paragraph block will still have the caret blinking inside it even though a different block might be selected. This seems to be an issue with how the accessibility flag is implemented on ReactNative, in combination with the Android OS underneath it. In particular, setting a block to `accessible=true` seems to cause a request for focus. Other Android OS versions (at least 8.0 and 9.0) don't seem to exhibit that behavior.

Update release notes:
